pontoon/base/views.py

```diff
@@ -168,111 +166,137 @@ def authors_and_time_range(request, locale, slug, part):
     }, safe=False)
 
 
+def _get_entities_list(locale, project, form):
+    """Return a specific list of entities, as defined by the `entity_ids` field of the form.
+
+    This is used for batch editing.
+    """
+    entities = (
+        Entity.objects.filter(pk__in=form.cleaned_data['entity_ids'])
+        .prefetch_resources_translations(locale)
+        .distinct()
+        .order_by('order')
+    )
+
+    return JsonResponse({
+        'entities': Entity.map_entities(locale, entities),
+        'stats': TranslatedResource.objects.stats(
+            project, form.cleaned_data['paths'], locale
+        ),
+    }, safe=False)
+
+
+def _get_all_entities(locale, project, form, entities):
+    """Return entities without pagination.
+
+    This is used by the in-context mode of the Translate page.
+    """
+    has_next = False
+    entities_to_map = Entity.for_project_locale(
+        project,
+        locale,
+        paths=form.cleaned_data['paths'],
+        exclude_entities=form.cleaned_data['exclude_entities']
+    )
+    visible_entities = entities.values_list('pk', flat=True)
+
+    return JsonResponse({
+        'entities': Entity.map_entities(locale, entities_to_map, visible_entities),
+        'has_next': has_next,
+        'stats': TranslatedResource.objects.stats(
+            project, form.cleaned_data['paths'], locale
+        ),
+    }, safe=False)
+
+
+def _get_paginated_entities(locale, project, form, entities):
+    """Return a paginated list of entities.
+
+    This is used by the regular mode of the Translate page.
+    """
+    paginator = Paginator(entities, form.cleaned_data['limit'])
+
+    try:
+        entities_page = paginator.page(1)
+    except EmptyPage:
+        return JsonResponse({
+            'has_next': False,
+            'stats': {},
+            'authors': [],
+            'counts_per_minute': [],
+        })
+
+    has_next = entities_page.has_next()
+    entities_to_map = entities_page.object_list
+
+    # If requested entity not on the first page
+    if form.cleaned_data['entity']:
+        entity_pk = form.cleaned_data['entity']
+
+        # TODO: entities_to_map.values_list() doesn't return entities from selected page
+        if entity_pk not in [e.pk for e in entities_to_map]:
+            if entity_pk in entities.values_list('pk', flat=True):
+                entities_to_map = list(entities_to_map) + list(entities.filter(pk=entity_pk))
+
+    return JsonResponse({
+        'entities': Entity.map_entities(locale, entities_to_map, []),
+        'has_next': has_next,
+        'stats': TranslatedResource.objects.stats(
+            project, form.cleaned_data['paths'], locale
+        ),
+    }, safe=False)
+
```

Comment:
I wonder if there's a better place to put these `_get_entities*` functions in. I don't have an idea though. :)

Reply:
Yeah, ultimately I want the ``translate`` page to live in a different app. It souldn't be in ``base``. It's a bit out of the scope of this, plus if we start rewriting that page, I'll do it in a different app, and in the end we'll be able to remove all of this from ``base``.